proposals/0000-add-sequence-based-init-and-merge-to-dictionary.md

```diff
@@ -0,0 +1,180 @@
+        var mutableDictionary = cumulative        // making a copy every time?
+        mutableDictionary[keyValue.0] = keyValue.1
+        return mutableDictionary
+    }
+
+Beyond initialization, `Array` and `Set` both also provide a method to add a new block of elements to an existing collection. `Array` provides this via `appendContentsOf(_:)` for the common appending case or `replaceRange(_:with:)` for general inserting or replacing, while the unordered `Set` type lets you pass any sequence to `unionInPlace(_:)` to add elements to an existing set.
+
+Once again, `Dictionary` has no corresponding API -- looping and adding elements one at a time as shown above is the only way to merge new elements into an existing dictionary.
+
+
+## Proposed solution
+
+### New initializer
+
+The proposed solution would add a pair of initializers to `Dictionary` that accept any sequence of `(Key, Value)` tuple pairs:
+
+    init<S: SequenceType where S.Generator.Element == (Key, Value)>(_ sequence: S)
+    init<S: SequenceType where S.Generator.Element == (Key, Value)>(_ sequence: S, @noescape combine: (Key, Value, Value) throws -> Value) rethrows
+
+Instead of the techniques for recovering a `Dictionary` instance shown above, the proposed initializer would allow a much cleaner syntax:
+
+    let viaProposed = Dictionary(evenOnly)
+
+Moreover, the new initializers would allow for some convenient uses that aren't currently possible.
+
+- Initializing from an array of tuples: 
+
+        let dictFromArray = Dictionary([("a", 1), ("b", 2), ("c", 3), ("d", 4)])
+
+- Initializing from a `DictionaryLiteral` (the type, not an actual literal): 
+
+        let literal: DictionaryLiteral = ["a": 1, "b": 2, "c": 3, "d": 4]
+        let dictFromDL = Dictionary(literal)
+
+- Swapping keys and values of an existing dictionary:
+
+        let reversedDict = Dictionary(dictFromDL.map { ($1, $0) })
+        // [2: "b", 4: "d", 1: "a", 3: "c"]
+
+- Converting an array to an indexed dictionary (popular on the thread):
+
+        let names = ["Cagney", "Lacey", "Bensen"]
+        let dict = Dictionary(names.enumerate().map { (i, val) in (i + 1, val) })
+        // [2: "Lacey", 3: "Bensen", 1: "Cagney"]
+
+- Initializing from a pair of zipped sequences (examples abound): 
+
+        let letters = "abcdefghij".characters.lazy.map { String($0) }
+        let dictFromZip = Dictionary(zip(letters, 1...10))
+        // ["b": 2, "a": 1, "i": 9, "j": 10, "c": 3, "e": 5, "f": 6, "g": 7, "d": 4, "h": 8]
+        
+    That last one might feel familiar to Cocoa developers accustomed to `dictionaryWithObjects:forKeys:`.
+
+#### Handling duplicate keys
+
+Creating a `Dictionary` from a dictional literal currently checks the keys for uniqueness, trapping on a duplicate. This behavior makes sense for literals, but would be cumbersome and unfriendly when initializing with a sequence. Therefore, this proposal includes a method for gracefully handling duplicate keys.
```

Comment:
I'm modifying the proposal to convert the simple `init` to be a failable `init?` and the merging `init` to use a `marging` label for the sequence: `init(merging:combine:)`. The new APIs under the revised proposal would be:

```swift
init?<S: SequenceType where S.Generator.Element == (Key, Value)>(
    _ sequence: S)
    
init<S: SequenceType where S.Generator.Element == (Key, Value)>(
    merging sequence: S, 
    @noescape combine: (Value, Value) throws -> Value
    ) rethrows
mutating func mergeContentsOf<S: SequenceType where S.Generator.Element == (Key, Value)>(
    _ sequence: S, 
    @noescape combine: (Value, Value) throws -> Value
    ) rethrows
func mergedWith<S: SequenceType where S.Generator.Element == (Key, Value)>(
    _ sequence: S, 
    @noescape combine: (Value, Value) throws -> Value) rethrows -> [Key: Value]
```
@gribozavr: I think this should address your concerns and let the caller decide what level of error checking they want. Thanks again for your comments, they really helped me clarify the difference between the two kinds of functionality requested here.
@NachoSoto: With this revision you would use the merging init to ignore duplicate keys and choose specifically which value to take. How does this look?
```swift
let dupes = [("a", 1), ("b", 2), ("a", 3), ("b", 4)]
let firstValueDictionary = Dictionary(merging: dupes) { (first, _) in first }
// ["b": 2, "a": 1]
let lastValueDictionary = Dictionary(merging: dupes) { (_, last) in last }
// ["b": 4, "a": 3]
```